### PR TITLE
[IDP-941] Added new analyzer rule to enforce DomainEvent name convention

### PR DESCRIPTION
**Pull request by heqianwang**
## Description of changes
Added new Roslyn Analyzer rule for enforcing the proper naming convention for DomainEvent names. It should follow the format of `com.{Product}.{DomainService}.{Action}` or `com.{Product}.{DomainService}.{Entity}.{Action}`

## Breaking changes
The new package may force users to correct their domain event names if it was incorrect.

## Additional checks
This is a follow-up to [the previous IDP-941 PR.](https://github.com/gsoft-inc/wl-domain-event-propagation/pull/109) which added the dotnet 8 sdk on the global.json.

- [x] Updated the documentation of the project to reflect the changes
- [x] Added new tests that cover the code changes

Tested with a local version of the package and verified that that the Roslyn analyzer rules were being applied correctly:
image

### CODE REVIEW

File: src/Workleap.DomainEventPropagation.Analyzers.Tests/Workleap.DomainEventPropagation.Analyzers.Tests.csproj [hunk 1]
```diff
@@ -15,7 +15,7 @@
     <PackageReference Include="Microsoft.CodeAnalysis.CSharp.Workspaces" Version="4.4.0" PrivateAssets="all" />
     <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.8.0" />
     <PackageReference Include="Polly" Version="8.2.1" />
-    <PackageReference Include="xunit" Version="2.6.6" />
+    <PackageReference Include="xunit" Version="2.4.2" />
```
Comment (meziantou) [inline src/Workleap.DomainEventPropagation.Analyzers.Tests/Workleap.DomainEventPropagation.Analyzers.Tests.csproj:18]: Can you update renovate.json to disable updates for this package?

File: src/Workleap.DomainEventPropagation.Analyzers/EventDomainAttributeUsageAnalyzer.cs [hunk 2]
```diff
@@ -26,9 +27,19 @@ public sealed class EventDomainAttributeUsageAnalyzer : DiagnosticAnalyzer
         defaultSeverity: DiagnosticSeverity.Warning,
         isEnabledByDefault: true,
         helpLinkUri: RuleIdentifiers.HelpUri);
+    
+    internal static readonly DiagnosticDescriptor FollowNamingConventionAttribute = new DiagnosticDescriptor(
+        id: RuleIdentifiers.FollowNamingConventionAttributeValue,
+        title: "Follow naming convention in attribute",
```
Comment (meziantou) [inline src/Workleap.DomainEventPropagation.Analyzers/EventDomainAttributeUsageAnalyzer.cs:33]:
```suggestion
        title: "Follow naming convention in DomainEvent attribute",
```

File: src/Workleap.DomainEventPropagation.Analyzers/EventDomainAttributeUsageAnalyzer.cs [hunk 3]
```diff
@@ -26,9 +27,19 @@ public sealed class EventDomainAttributeUsageAnalyzer : DiagnosticAnalyzer
...
+        messageFormat: "Follow naming convention in attribute: {0}",
```
Comment (meziantou) [inline src/Workleap.DomainEventPropagation.Analyzers/EventDomainAttributeUsageAnalyzer.cs:34]:
```suggestion
        messageFormat: "Follow naming convention in DomainEvent attribute: {0}",
```

File: src/Workleap.DomainEventPropagation.Analyzers/EventDomainAttributeUsageAnalyzer.cs [hunk 4]
```diff
@@ -26,9 +27,19 @@ public sealed class EventDomainAttributeUsageAnalyzer : DiagnosticAnalyzer
...
+        description: "Follow naming convention in attribute, the DomainEvent should follow the reverse dns format com.{Product}.{DomainService}.{Action} or com.{Product}.{DomainService}.{Entity}.{Action} all in lowercase.",
```
Comment (meziantou) [inline src/Workleap.DomainEventPropagation.Analyzers/EventDomainAttributeUsageAnalyzer.cs:35]:
```suggestion
        description: "Follow naming convention in attribute, the DomainEvent must be all in lowercase and follow the reverse dns format com.{Product}.{DomainService}.{Action} or com.{Product}.{DomainService}.{Entity}.{Action}.",
```

**Comment by PrincessMadMath** [comment]
Next time I think this PR could have been split into 2 PR:
1. Updating the SDK version + using the new string formatting in existing test
2. Add the new test

File: src/Workleap.DomainEventPropagation.Analyzers/EventDomainAttributeUsageAnalyzer.cs [hunk 5]
```diff
@@ -89,12 +101,77 @@ public void AnalyzeOperationInvocation(SymbolAnalysisContext context)
...
+
+            if (listOfUrlComponents.Any(s => s.Length == 0))
+            {
+                invalidNameReason = "The domain event name should not have empty segments.";
+                return false;
+            }
+            
+            if (!AllowedProductNames.Contains(listOfUrlComponents[1]))
+            {
+                invalidNameReason = "The domain event name product field should be part of the product list.";
```
Comment (meziantou) [inline src/Workleap.DomainEventPropagation.Analyzers/EventDomainAttributeUsageAnalyzer.cs:161]: You can add the list of allowed products to the error message

**Review by asimmon — APPROVED** [review]

**Pull request merged**